core/drivers/crypto/caam/mp/caam_mp.c

```diff
@@ -0,0 +1,405 @@
+	if (!msg || !msg_size || !sig || !sig_size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (*sig_size < 2 * mp_privdata.sec_size) {
+		*sig_size = 2 * mp_privdata.sec_size;
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+
+	if (*msg_size > MP_SIGN_MAX_MSG_SIZE)
+		return TEE_ERROR_EXCESS_DATA;
+
+	/* Re-allocate the message to a cache-aligned buffer */
+	aligned_msg = caam_alloc(*msg_size);
+	if (!aligned_msg) {
+		MP_TRACE("Message reallocation error");
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto exit_mpsign;
+	}
+	memcpy(aligned_msg, msg, *msg_size);
+
+	/*
+	 * Allocate the hash buffer of the Message + MPMR payload
+	 * Note: Hash is not retrieve, hence no need to do cache
+	 * maintenance
+	 */
+	retstatus = caam_alloc_align_buf(&hash, TEE_MAX_HASH_SIZE);
+	if (retstatus != CAAM_NO_ERROR) {
+		MP_TRACE("Hash allocation error");
+		ret = caam_status_to_tee_result(retstatus);
+		goto exit_mpsign;
+	}
+
+	/*
+	 * Re-allocate the signature result buffer with a maximum size
+	 * of the roundup to 16 bytes of the secure size in bytes if
+	 * the signature buffer is not aligned or too short.
+	 *
+	 *  - 1st Part: size_sec
+	 *  - 2nd Part: size_sec roundup to 16 bytes
+	 */
+	sign_len = ROUNDUP(mp_privdata.sec_size, 16) + mp_privdata.sec_size;
+
+	ret = caam_dmaobj_output_sgtbuf(&sign_c, sig, *sig_size, sign_len);
+	if (ret)
+		goto exit_mpsign;
+
+	if (sign_c.sgtbuf.sgt_type)
+		pdb_sgt_flags |= PDB_SGT_MP_SIGN_C;
+
+	/* Prepare the 2nd Part of the signature. Derived from sign_c */
+	ret = caam_dmaobj_derive_sgtbuf(&sign_d, &sign_c, mp_privdata.sec_size,
+					ROUNDUP(mp_privdata.sec_size, 16));
+	if (ret)
+		goto exit_mpsign;
+
+	if (sign_d.sgtbuf.sgt_type)
+		pdb_sgt_flags |= PDB_SGT_MP_SIGN_D;
+
+	ret = caam_dmaobj_input_sgtbuf(&msg_input, aligned_msg, *msg_size);
+	if (ret)
+		goto exit_mpsign;
+
+	if (msg_input.sgtbuf.sgt_type)
+		pdb_sgt_flags |= PDB_SGT_MP_SIGN_MSG;
+
+	desc = caam_calloc_desc(MP_SIGN_DESC_ENTRIES);
+	if (!desc) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto exit_mpsign;
+	}
+
+	caam_desc_init(desc);
+	caam_desc_add_word(desc, DESC_HEADER(0));
+	caam_desc_add_word(desc,
+			   PROT_MP_CURVE(mp_privdata.curve) | pdb_sgt_flags);
+	caam_desc_add_ptr(desc, msg_input.sgtbuf.paddr);
+	caam_desc_add_ptr(desc, hash.paddr);
+	caam_desc_add_ptr(desc, sign_c.sgtbuf.paddr);
+	caam_desc_add_ptr(desc, sign_d.sgtbuf.paddr);
+	caam_desc_add_word(desc, msg_input.sgtbuf.length);
+	caam_desc_add_word(desc, MPSIGN_OP);
+
+	desclen = caam_desc_get_len(desc);
+	caam_desc_update_hdr(desc, DESC_HEADER_IDX(desclen, desclen - 1));
+
+	MP_DUMPDESC(desc);
+
+	caam_dmaobj_cache_push(&msg_input);
+	caam_dmaobj_cache_push(&sign_c);
+
+	jobctx.desc = desc;
+	retstatus = caam_jr_enqueue(&jobctx, NULL);
+
+	if (retstatus == CAAM_NO_ERROR) {
+		sign_c.orig.length = 2 * mp_privdata.sec_size;
+		*sig_size = caam_dmaobj_copy_to_orig(&sign_c);
+
+		MP_DUMPBUF("MP Signature", sdata->signature.data,
+			   sdata->signature.length);
+
+		ret = caam_status_to_tee_result(retstatus);
```

Comment:
likely `ret = TEE_SUCCESS;` prefer the explicit way?

Reply:
I prefer `ret = caam_status_to_tee_result(retstatus);`